src/libsyntax/codemap.rs

```diff
@@ -807,6 +807,70 @@ impl CodeMap {
                         hi.col.to_usize() + 1)).to_string()
     }
 
+    pub fn span_to_expanded_string(&self, sp: Span) -> String {
+        let src_str = self.span_to_string(sp);
+        let src_snip = self.span_to_snippet(sp).unwrap_or(
+            "Snippet unavailable (span malformed or source missing)".to_string());
+        if sp.expn_id == NO_EXPANSION || sp.expn_id == COMMAND_LINE_EXPN {
+            return format!("{}\n{}", src_snip, src_str);
+        }
+
+        let mut strings = vec![src_str.into_boxed_str(), src_snip.into_boxed_str()];
+        let mut cur_span = sp;
+        while cur_span.expn_id != NO_EXPANSION && cur_span.expn_id != COMMAND_LINE_EXPN {
+            let old_span = cur_span;
+            let new_span = self.with_expn_info(old_span.expn_id,
+                                              |ei| ei.map(|ei| ei.call_site));
+
+            //No info attached to src span, return
+            if let None = new_span {
+                return format!("{}\n{}", &*strings[1], &*strings[0]);
+            }
+            let new_span = new_span.unwrap();
+
+            //This compacts recursive macros down into a single entry,
+            //By only pushing sequential spans with different callees
+            let old_fmt = self.with_expn_info(
```

Comment:
It shouldn't squash the trace - it just only prints one entry for that macro. It will still print all spans before and after. I'll clarify the comment